### PR TITLE
СhangeLogImplementation

### PR DESCRIPTION
**Pull request by YariLAN**

### CODE REVIEW

File: src/DashboardService.Data/ChangeLogRepository.cs [hunk 1]
```diff
@@ -19,28 +21,94 @@ public ChangeLogRepository(IDataProvider provider)
...
+      changeLogs = filter.IsAscendingSortByCreatedAtUtc.Value
+        ? changeLogs.OrderBy(db => db.CreatedAtUtc)
+        : changeLogs.OrderByDescending(db => db.CreatedAtUtc);
+    }
+
+    if (filter.IsAscendingSortByEntityName.HasValue)
+    {
+      changeLogs = filter.IsAscendingSortByEntityName.Value
+        ? changeLogs.OrderBy(db => db.EntityName)
+                    .ThenByDescending(db => db.CreatedAtUtc)
```
Comment (VLiamin) [inline src/DashboardService.Data/ChangeLogRepository.cs]: tabs

File: src/DashboardService.Data/ChangeLogRepository.cs [hunk 2]
```diff
@@ -19,28 +21,94 @@ public ChangeLogRepository(IDataProvider provider)
...
+    await _provider.SaveAsync();
+
+    return true;
   }
 
-  public Task<DbChangeLog> GetAsync(Guid id, CancellationToken ct)
+  public async Task<DbChangeLog> GetAsync(Guid id, CancellationToken ct)
   {
-    throw new NotImplementedException();
+    IQueryable<DbChangeLog> changeLog = _provider.ChangeLogs.AsNoTracking();
```
Comment (VLiamin) [inline src/DashboardService.Data/ChangeLogRepository.cs]: You can remove this variable

Comment (VLiamin) [reply src/DashboardService.Data/ChangeLogRepository.cs]: and use ct

File: src/DashboardService.Models.Dto/Requests/ChangeLog/Filter/GetChangeLogsFilter.cs [hunk 3]
```diff
@@ -9,12 +9,15 @@ public record GetChangeLogsFilter : BaseFindFilter
...
+  public bool? IsAscendingSortByEntityName { get; set; }
 
-  [FromQuery(Name = "isAscendingSort")]
-  public bool? IsAscendingSort { get; set; }
+  [FromQuery(Name = "entityNameIncludeSubstring")]
+  public string EntityNameIncludeSubstring { get; set; }
 
-  [FromQuery(Name = "nameIncludeSubstring")]
-  public string NameIncludeSubstring { get; set; }
+  [FromQuery(Name = "PropertyNameIncludeSubstring")]
```
Comment (VLiamin) [inline src/DashboardService.Models.Dto/Requests/ChangeLog/Filter/GetChangeLogsFilter.cs]: all atr. names in camelCase(entityNameIncludeSubstring, isAscendingSortByEntityName). This in PascalCase.

File: src/DashboardService.Data/ChangeLogRepository.cs [hunk 4]
```diff
@@ -19,28 +21,94 @@ public ChangeLogRepository(IDataProvider provider)
...
+    }
+
+    if (filter.IsAscendingSortByEntityName.HasValue)
+    {
+      changeLogs = filter.IsAscendingSortByEntityName.Value
+        ? changeLogs.OrderBy(db => db.EntityName)
+                    .ThenByDescending(db => db.CreatedAtUtc)
+
+        : changeLogs.OrderByDescending(db => db.EntityName)
+                    .ThenByDescending(db => db.CreatedAtUtc);
```
Comment (VLiamin) [inline src/DashboardService.Data/ChangeLogRepository.cs]: Can filter.IsAscendingSortByCreatedAtUtc be true(OrderBy// not Descending//)?

Comment (YariLAN) [reply src/DashboardService.Data/ChangeLogRepository.cs]: Why not Descending? Usually need last changes?

File: src/DashboardService.Models.Dto/Requests/ChangeLog/Filter/GetChangeLogsFilter.cs (hunk 3 above)
Comment (YariLAN) [reply src/DashboardService.Models.Dto/Requests/ChangeLog/Filter/GetChangeLogsFilter.cs]: ++

File: src/DashboardService.Data/ChangeLogRepository.cs (hunk 2 above)
Comment (YariLAN) [reply src/DashboardService.Data/ChangeLogRepository.cs]: ++

File: src/DashboardService.Data/ChangeLogRepository.cs (hunk 1 above)
Comment (YariLAN) [reply src/DashboardService.Data/ChangeLogRepository.cs]: ++

File: src/DashboardService.Data/ChangeLogRepository.cs [hunk 5]
```diff
@@ -19,28 +21,102 @@ public ChangeLogRepository(IDataProvider provider)
...
+    {
+      changeLogs = filter.IsAscendingSortByCreatedAtUtc.Value
+        ? changeLogs.OrderBy(db => db.CreatedAtUtc)
+        : changeLogs.OrderByDescending(db => db.CreatedAtUtc);
+    }
+
+    if (filter.IsAscendingSortByEntityName.HasValue && !filter.IsAscendingSortByCreatedAtUtc.HasValue)
+    {
+      changeLogs = filter.IsAscendingSortByEntityName.Value
+              ? changeLogs.OrderBy(x => x.EntityName)
```
Comment (VLiamin) [inline src/DashboardService.Data/ChangeLogRepository.cs]: tabs

Comment (YariLAN) [reply src/DashboardService.Data/ChangeLogRepository.cs]: ++

File: src/DashboardService.Business/ChangeLog/GetChangeLogCommand.cs [hunk 6]
```diff
@@ -27,8 +28,16 @@ public class GetChangeLogCommand : IGetChangeLogCommand
     _responseCreator = responseCreator;
   }
 
-  public Task<OperationResultResponse<ChangeLogInfo>> ExecuteAsync(Guid id, CancellationToken ct)
+  public async Task<OperationResultResponse<ChangeLogInfo>> ExecuteAsync(Guid id, CancellationToken ct)
```
Comment (SolovyovDmitry) [inline src/DashboardService.Business/ChangeLog/GetChangeLogCommand.cs]: specify: id of what?

File: src/DashboardService.Business/ChangeLog/GetChangeLogsCommand.cs [hunk 7]
```diff
@@ -13,21 +15,32 @@ namespace LT.DigitalOffice.DashboardService.Business.ChangeLog;
...
     _changeLogInfoMapper = changeLogInfoMapper;
   }
 
-  public Task<FindResultResponse<ChangeLogInfo>> ExecuteAsync(GetChangeLogsFilter filter, CancellationToken ct)
+  public async Task<FindResultResponse<ChangeLogInfo>> ExecuteAsync(GetChangeLogsFilter filter, CancellationToken ct)
   {
-    throw new System.NotImplementedException();
+    (List<DbChangeLog> dbChanges, int totalCount) = await _repository.GetChangeLogsAsync(filter, ct);
+
+    if (dbChanges is null || !dbChanges.Any())
```
Comment (SolovyovDmitry) [inline src/DashboardService.Business/ChangeLog/GetChangeLogsCommand.cs]: prefer .Count to .Any()

File: src/DashboardService.Data/Interfaces/IChangeLogRepository.cs [hunk 8]
```diff
@@ -12,9 +12,9 @@ namespace LT.DigitalOffice.DashboardService.Data.Interfaces;
 [AutoInject]
 public interface IChangeLogRepository
 {
-  Task<Guid?> CreateAsync(DbChangeLog board, CancellationToken ct);
+  Task<Guid?> CreateAsync(DbChangeLog board);
```
Comment (SolovyovDmitry) [inline src/DashboardService.Data/Interfaces/IChangeLogRepository.cs]: board?

File: src/DashboardService.Data/ChangeLogRepository.cs [hunk 9]
```diff
@@ -19,28 +21,102 @@ public ChangeLogRepository(IDataProvider provider)
...
+    if (changeLog is null)
+    {
+      return false;
+    }
+
+    _provider.ChangeLogs.Remove(changeLog);
+    await _provider.SaveAsync();
+
+    return true;
   }
```
Comment (SolovyovDmitry) [inline src/DashboardService.Data/ChangeLogRepository.cs]: don't think this is required

Comment (YariLAN) [reply src/DashboardService.Data/ChangeLogRepository.cs]: We won't delete the logs?

File: src/DashboardService.Data/Interfaces/IChangeLogRepository.cs (hunk 8 above)
Comment (YariLAN) [reply src/DashboardService.Data/Interfaces/IChangeLogRepository.cs]: ++

File: src/DashboardService.Business/ChangeLog/GetChangeLogsCommand.cs (hunk 7 above)
Comment (YariLAN) [reply src/DashboardService.Business/ChangeLog/GetChangeLogsCommand.cs]: ++

File: src/DashboardService.Business/ChangeLog/GetChangeLogCommand.cs (hunk 6 above)
Comment (YariLAN) [reply src/DashboardService.Business/ChangeLog/GetChangeLogCommand.cs]: ++

File: src/DashboardService.Business/ChangeLog/GetChangeLogsCommand.cs [hunk 10]
```diff
@@ -29,7 +29,7 @@ public async Task<FindResultResponse<ChangeLogInfo>> ExecuteAsync(GetChangeLogsF
   {
     (List<DbChangeLog> dbChanges, int totalCount) = await _repository.GetChangeLogsAsync(filter, ct);
 
-    if (dbChanges is null || !dbChanges.Any())
+    if (dbChanges is null || dbChanges.Count() == 0)
```
Comment (SolovyovDmitry) [inline src/DashboardService.Business/ChangeLog/GetChangeLogsCommand.cs]: count without (). It makes difference

File: src/DashboardService.Business/ChangeLog/GetChangeLogCommand.cs [hunk 11]
```diff
@@ -28,9 +28,9 @@ public class GetChangeLogCommand : IGetChangeLogCommand
     _responseCreator = responseCreator;
   }
 
-  public async Task<OperationResultResponse<ChangeLogInfo>> ExecuteAsync(Guid id, CancellationToken ct)
+  public async Task<OperationResultResponse<ChangeLogInfo>> ExecuteAsync(Guid changeLogId, CancellationToken ct)
```
Comment (SolovyovDmitry) [inline src/DashboardService.Business/ChangeLog/GetChangeLogCommand.cs]: I don't see any use cases for it. The only case I see is getting all change logs by parent entity id

File: src/DashboardService.Data/ChangeLogRepository.cs (hunk 2 above)
Comment (SolovyovDmitry) [reply src/DashboardService.Data/ChangeLogRepository.cs]: not fixed

File: src/DashboardService.Data/ChangeLogRepository.cs (hunk 9 above)
Comment (SolovyovDmitry) [reply src/DashboardService.Data/ChangeLogRepository.cs]: I guess we will but not by single change log id but all of them by parent entity id

File: src/DashboardService.Data/ChangeLogRepository.cs (hunk 5 above)
Comment (SolovyovDmitry) [reply src/DashboardService.Data/ChangeLogRepository.cs]: not fixed

File: src/DashboardService.Data/ChangeLogRepository.cs (hunk 1 above)
Comment (SolovyovDmitry) [reply src/DashboardService.Data/ChangeLogRepository.cs]: not fixed

File: src/DashboardService.Data/ChangeLogRepository.cs [hunk 12]
```diff
@@ -19,28 +21,102 @@ public ChangeLogRepository(IDataProvider provider)
...
   {
-    throw new NotImplementedException();
+    DbChangeLog changeLog = await _provider.ChangeLogs.FirstOrDefaultAsync(x => x.Id == changeLogId, ct);
+
+    if (changeLog is null || request is null)
+    {
+      return false;
+    }
+
+    request.ApplyTo(changeLog);
```
Comment (VLiamin) [inline src/DashboardService.Data/ChangeLogRepository.cs:42]: Edit ModifiedBy?

Comment (YariLAN) [reply src/DashboardService.Data/ChangeLogRepository.cs:42]: It doesn't exist in Change Log

Comment (YariLAN) [reply src/DashboardService.Data/ChangeLogRepository.cs:42]: It exists in Board

File: src/DashboardService.Business/ChangeLog/GetChangeLogCommand.cs (hunk 11 above)
Comment (YariLAN) [reply src/DashboardService.Business/ChangeLog/GetChangeLogCommand.cs]: ++

File: src/DashboardService.Business/ChangeLog/GetChangeLogsCommand.cs (hunk 10 above)
Comment (YariLAN) [reply src/DashboardService.Business/ChangeLog/GetChangeLogsCommand.cs]: ++

File: src/DashboardService.Data/ChangeLogRepository.cs (hunk 9 above)
Comment (YariLAN) [reply src/DashboardService.Data/ChangeLogRepository.cs]: ++

File: src/DashboardService.Data/ChangeLogRepository.cs (hunk 5 above)
Comment (YariLAN) [reply src/DashboardService.Data/ChangeLogRepository.cs]: fixed

File: src/DashboardService.Data/ChangeLogRepository.cs (hunk 2 above)
Comment (YariLAN) [reply src/DashboardService.Data/ChangeLogRepository.cs]: fixed

File: src/DashboardService.Data/ChangeLogRepository.cs (hunk 1 above)
Comment (YariLAN) [reply src/DashboardService.Data/ChangeLogRepository.cs]: fixed

**Review by VLiamin — APPROVED** [review]

File: src/DashboardService.Business/ChangeLog/GetChangeLogsCommand.cs [hunk 13]
```diff
@@ -13,21 +14,32 @@ namespace LT.DigitalOffice.DashboardService.Business.ChangeLog;
...
-    throw new System.NotImplementedException();
+    (List<DbChangeLog> dbChanges, int totalCount) = await _repository.GetChangeLogsAsync(filter, ct);
+
+    if (dbChanges is null || dbChanges.Count == 0)
+    {
+      return new()
+      {
+        Body = new List<ChangeLogInfo>(),
+      };
+    }
```
Comment (SolovyovDmitry) [inline src/DashboardService.Business/ChangeLog/GetChangeLogsCommand.cs]: unnecessary

**Review by SolovyovDmitry — APPROVED** [review]

File: src/DashboardService.Business/ChangeLog/GetChangeLogsCommand.cs (hunk 13 above)
Comment (YariLAN) [reply src/DashboardService.Business/ChangeLog/GetChangeLogsCommand.cs]: ++

**Pull request merged**